### PR TITLE
BEP341 & BEP402

### PR DESCRIPTION
**Pull request by yoshidan**
https://github.com/bnb-chain/BEPs/blob/master/BEPs/BEP-341.md
https://github.com/bnb-chain/BEPs/blob/master/BEPs/BEP-402.md

### CODE REVIEW

File: light-client/src/header/eth_header.rs [hunk 1]
```diff
@@ -105,6 +113,32 @@ impl ETHHeader {
         stream.append(&self.extra_data[..self.extra_data.len() - EXTRA_SEAL].to_vec());
         stream.append(&self.mix_digest);
         stream.append(&self.nonce);
+        if let Some(parent_beacon_root) = &self.parent_beacon_root {
```
Comment (yoshidan) [inline light-client/src/header/eth_header.rs:116]: Calculation of the seal hash according to BEP-402

File: light-client/src/header/eth_header.rs [hunk 2]
```diff
@@ -236,45 +270,44 @@ impl ETHHeader {
...
-    pub fn is_epoch(&self) -> bool {
-        self.number % BLOCKS_PER_EPOCH == 0
+    let num = extra_data[EXTRA_VANITY] as usize;
+    if num == 0 || extra_data.len() <= EXTRA_VANITY + EXTRA_SEAL + num * VALIDATOR_BYTES_LENGTH {
+        return Err(Error::UnexpectedExtraDataLength(extra_data.len()));
     }
+    let start = EXTRA_VANITY + VALIDATOR_NUM_SIZE;
+    let end = start + num * VALIDATOR_BYTES_LENGTH;
+    let turn_length = extra_data[end];
+    if !(turn_length == 1 || (3..=9).contains(&turn_length)) {
```
Comment (yoshidan) [inline light-client/src/header/eth_header.rs:301]: Only 1 and 3~9 are allowed according to 424-governable-number-of-consecutive-blocks

File: light-client/src/header/epoch.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct Epoch {
+    validator_set: ValidatorSet,
+    turn_length: u8,
+    hash: Hash,
+}
+
+impl Epoch {
+    pub fn new(validator_set: ValidatorSet, turn_length: u8) -> Self {
+        let seed = [[turn_length].as_slice(), validator_set.hash.as_slice()].concat();
```
Comment (yoshidan) [inline light-client/src/header/epoch.rs:15]: Save this hash as a validator hash in the cons state.

File: light-client/src/fixture/localnet.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+use crate::header::eth_header::ETHHeader;
+use crate::header::eth_headers::ETHHeaders;
+
+use crate::fixture::{decode_header, Network, UpdateClientEpochInput, UpdateClientNonEpochInput};
+use crate::misc::{Address, ChainId};
+
+pub struct Localnet;
+
+impl Network for Localnet {
```
Comment (yoshidan) [inline light-client/src/fixture/localnet.rs:13]: Test data can be prepared for each network for future HF support. Currently, Bohr is only operating on the local net, so the test data is for the local net only.
We will increase the corresponding test data once it is deployed on the test net and main net.

File: light-client/src/header/mod.rs [hunk 5]
```diff
@@ -197,24 +198,32 @@ impl TryFrom<RawHeader> for Header {
...
             headers
                 .target
-                .get_validator_bytes()
-                .ok_or_else(|| Error::MissingValidatorInEpochBlock(headers.target.number))?
+                .epoch
+                .clone()
+                .ok_or_else(|| Error::MissingEpochInfoInEpochBlock(headers.target.number))?
         } else {
-            value.current_validators
+            //TODO validate turn term
```
[the remark this answers is not in the document]
Comment (yoshidan) [reply light-client/src/header/mod.rs]: I added the validation.
https://github.com/datachainlab/parlia-elc/pull/52/commits/bbd0938ed28a8513dc5052796bead0321d5bc4bf#diff-98593905abb552727b9a6265008748ad27c594ea86c62199ff93c4e6a9d95984R217

**Review by 3100 — APPROVED** [review]

**Pull request merged**